Allows user to specify search terms and adds Album art from Last.fm

g33kyaditya opened the pull request:
The current Album Art backend is Google (default) and Last.fm which is available to the user by selecting an optional argument ~~which although works fine is not enough. Hence it is WIP and I will add more backends like MusicBrainz once I know that this is the way to go about it.~~
MusicBrainz uses coverartarchive.org as backend which is terrible. Last.fm has a much richer database and I'm sure if Last.fm doesn't have the album art, MusicBrainz won't have it too.

~~Additionally, I'm not sure whether to delete the downloaded images after they have been embedded to the mp3, so I've kept them for now.~~

An additional command line argument for specifying the search terms as a string has also been provided

g33kyaditya committed:
Allows user to specify search terms and adds Album art from Last.fm

g33kyaditya committed:
Finalized Last.fm as album art backend

ashishmadeti commented on download_from_csv.py:
```diff
@@ -77,6 +80,37 @@ def download_finish(d):
+    album = song['album']
+    track = song['name']
+    #Last.fm
+    API_KEY = '03f624c25bd2ffdbad319226716efe66'
```
You should never commit API keys to your code. It should either be an environment variable or a command line argument.

ashishmadeti commented on download_from_csv.py:
```diff
@@ -77,6 +80,37 @@ def download_finish(d):
+
+    image_path = folder + '/' + track + ' - ' + artist + '.png'
+    try:
+        urlretrieve(image_download, image_path)
```
Use `requests` library to fetch the image. Do not save the image to the file as you will have to read it from file anyway. Instead, return the image stream from your `download_album_art` method.
That way you can directly set the album art (without the need of a temporary file) in the main script

ashishmadeti commented on download_from_csv.py:
```diff
@@ -5,20 +5,23 @@
                     action="store_true")
 parser.add_argument('--skip', help="number of songs to skip from the start of csv",
                     type=int)
+parser.add_argument('-q', '--query', help="enter the search query", type=str)
```
Change the tone of the help string.
Help strings generally describe **what the argument** does instead of telling the user **what to do**. (You can confirm this by looking at the myriad of man pages available).
Ex: **number of songs to skip** instead of **enter the number of songs you want to skip**

ashishmadeti commented on download_from_csv.py:
```diff
@@ -77,6 +80,37 @@ def download_finish(d):
         print '\x1b[1A\x1b[2K'
         print "\x1b[1A[\033[93mConverting\033[00m] %s" % d['filename']
 
+def download_album_art(song):
```
It will become ugly when you would like to add more album-art fetchers. So you can do the following:

1. Move this functionality to a different module and import it here
2. Make a several functions, one for each album art backend. like `fetch_from_lastfm`, `fetch_from_google` etc.
3. Create a function which will act as an entry point for all the different backends (See point 2) and will take different parameters depending on the backend the user want to use (You can use [kwargs](http://www.saltycrane.com/blog/2008/01/how-to-use-args-and-kwargs-in-python/))

As users wouldn't bother to get their last.fm API key, add Google Images backend too and set it as default. Doing a query with `<Album name> cover art` gives very good results. Also add a command line argument to switch to last.fm backend (in which case your script will expect an environment variable with the API key)

ashishmadeti replied:
You can also think of some other better way to do it.

g33kyaditya committed:
Refactored album art fetcher, made google images default backend

g33kyaditya committed:
Removed unnecessary urllib import

g33kyaditya commented:
Fixed code and made changes according to last review. Added images.google.com as the default album art backend and gave user an option to switch to last.fm if needed.

ashishmadeti commented on album_art.py:
```diff
@@ -0,0 +1,84 @@
+        print 'No link found'
+        return
+    image_download = link.decode('unicode-escape')
+    #print 'Image download link is %s' %image_download
```
Cleanup

ashishmadeti commented on download_from_csv.py:
```diff
@@ -98,16 +101,35 @@ def download_finish(d):
+
+    image = None
+    if args.art == None:
+        image = album_art.fetch(song, backend='google')
```
Reduce the code in your if-else
The only difference between this line and L120 is `backend=<something>`

ashishmadeti commented on download_from_csv.py:
```diff
@@ -98,16 +101,35 @@ def download_finish(d):
+    image = None
+    if args.art == None:
+        image = album_art.fetch(song, backend='google')
+        if image != None:
```
`image` is same as `image != None`

Checkout https://docs.python.org/2.4/lib/truth.html

ashishmadeti commented on album_art.py:
```diff
@@ -0,0 +1,84 @@
+        print 'No images found for this album/track'
+        return
+    for image in reversed(images):
+        if image.get('size') == 'extralarge':       #The most suitable image size
```
Point of this if-else ladder?

g33kyaditya replied:
We are not sure if there is a extralarge for some image or not, so I download the best possible size among the available ones

ashishmadeti replied:
But you are doing the same thing in bodies of all the _ifs_. Might as well remove them, no?

ashishmadeti commented on album_art.py:
```diff
@@ -0,0 +1,84 @@
+    image_download = ''
+    images = []
+    try:
+        images = response.get('album', None).get('image', None)
```
This will result in an `AttributeError` when `response` does not contain the key _album_

ashishmadeti commented on album_art.py:
```diff
@@ -0,0 +1,84 @@
+    elif kwargs['backend'] == 'google':
+        image = fetch_from_google(song)
+    return image
+
```
Please fix PEP8 errors in all the your changes.

ashishmadeti replied:
There are pep8 plugins for almost all popular editors or you can install the `pep8` apt package and use it like `pep8 <your file name>` to check if your file conforms to pep8 guidelines.

Only acceptable error is `E501`

ashishmadeti commented on album_art.py:
```diff
@@ -0,0 +1,84 @@
+    tag = soup.find('div', {'class': 'rg_meta'})    #The first result
+    dictionary = json.loads(tag.string)
+    link = dictionary.get('ou', None)
+    if link == None:
```
`not link`

ashishmadeti replied:
Try to make your code more elegant.

ashishmadeti commented on album_art.py:
```diff
@@ -0,0 +1,84 @@
+
+    query = song['artist'] + song['name'] + 'cover art google play' # Art work can be of huge size so the 'google play' in query mostly results into smaller images
+    query = string.replace(query, ' ', '+')
+    url = 'https://www.google.com/search?site=&tbm=isch&source=hp&biw=1366&bih=649&q=%s&oq=%s&gs_l=img.3...857.12029.0.12205.36.14.1.20.21.0.301.1337.0j7j0j1.8.0....0...1ac.1.64.img..7.15.1346.0..0j0i24k1.22gw0wRgCfk'%(query, query)
```
Change it to most-minimal url that works.

ashishmadeti commented on album_art.py:
```diff
@@ -0,0 +1,84 @@
+import requests
+import os
+import string
```
No need for this. All string objects have `replace` member function.

ashishmadeti commented on album_art.py:
```diff
@@ -0,0 +1,84 @@
+def fetch_from_google(song):
+    ''' Downloads art from images.google.com '''
+
+    query = song['artist'] + song['name'] + 'cover art google play' # Art work can be of huge size so the 'google play' in query mostly results into smaller images
```
I don't know on what songs you tested this query on, but when I am testing this it is not showing good results.
Use `'%s %s' % (song['album'], song['artist'])`.

ashishmadeti replied:
Using `%` for concatenating string is more efficient than using `+` and also more readable.
Or even better use `' '.join(<array>)`

ashishmadeti commented on download_from_csv.py:
```diff
@@ -98,16 +101,35 @@ def download_finish(d):
         'outtmpl': folder + '/' + song['name'] + ' - ' + song['artist'] + '.%(ext)s'
     }
-    url = ' '.join([song['name'], song['artist'], 'audio', 'youtube'])
+    if args.query == None:
```
Reduce code in if-else.
Think: You are basically using `.join()` on an array. The array can change according to `args.query`

g33kyaditya committed:
Refactored code and followed PEP8 guidelines

g33kyaditya committed:
Changed gvsearch1 default query, removed unnecessary json import

g33kyaditya committed:
Cleanup
